File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/chat/MessageTask.kt
package jp.co.toukei.log.trustar.chat

import android.util.ArrayMap
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.core.Maybe
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.internal.disposables.ListCompositeDisposable
import io.reactivex.rxjava3.internal.functions.Functions
import jp.co.toukei.log.lib.addTo
import jp.co.toukei.log.lib.delayedRetry
import jp.co.toukei.log.lib.flatMapCompletable
import jp.co.toukei.log.lib.moveOrCopy
import jp.co.toukei.log.lib.observeOnIO
import jp.co.toukei.log.lib.orElseNull
import jp.co.toukei.log.lib.subscribeOnIO
import jp.co.toukei.log.lib.subscribeOrIgnore
import jp.co.toukei.log.trustar.Current
import jp.co.toukei.log.trustar.db.chat.Attachment
import jp.co.toukei.log.trustar.db.chat.ChatDB
import jp.co.toukei.log.trustar.db.chat.ChatMessagePendingDao
import jp.co.toukei.log.trustar.db.chat.entity.ChatMessagePending
import jp.co.toukei.log.trustar.repo.chat.ChatRepository
import jp.co.toukei.log.trustar.rest.model.FileKey
import jp.co.toukei.log.trustar.rest.model.FileKey2
import jp.co.toukei.log.trustar.toState
import jp.co.toukei.log.trustar.user.UserInfo
import okhttp3.MultipartBody
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.asRequestBody
import java.util.Collections
import java.util.concurrent.TimeUnit

private class FileRepository {

    fun msgUploadFile(filename: String, body: RequestBody): Single<FileKey> {
        return msgUploadFile2(filename, body).map { it.key }
    }

    fun msgUploadFile2(filename: String, body: RequestBody): Single<FileKey2> {
        val p = MultipartBody.Part.createFormData("buffer", filename, body)
        return Current.chatApi.uploadFile(p)
    }

    fun msgUploadImage(filename: String, body: RequestBody): Single<FileKey> {
        val p = MultipartBody.Part.createFormData("buffer", filename, body)
        return Current.chatApi.uploadImage(p)
    }

    fun msgUploadVideo(filename: String, body: RequestBody): Single<FileKey2> {
        return msgUploadFile2(filename, body)
    }
}

class MessageTask : Disposable {

    private val fileRepository = FileRepository()

    private val dc = ListCompositeDisposable()
    private val tasks = Collections.synchronizedMap(ArrayMap<String, Disposable>())

    fun observePendingChanges() {
        dc.clear()
        tasks.clear()
        val db = Current.chatDB.getInstance()
        db.chatMessagePendingDao().unsentOrSendingRoomIds()
            .subscribeOrIgnore { l ->
                l.forEach { roomTask(db, it.id) }
            }
            .addTo(dc)
    }

    //buggy?
    private fun roomTask(db: ChatDB, id: String): Boolean {
        val d = tasks[id]
        if (d != null && !d.isDisposed) return false
        val t = db.sendAndReceiveMessage(id)
            .doFinally {
                tasks.remove(id)?.let(dc::remove)
            }
            .subscribe(
                Functions.EMPTY_ACTION,
                Functions.emptyConsumer()
            )
        t.addTo(dc)
        tasks.put(id, t)?.let(dc::remove)
        return true
    }

    fun sendAndReceiveMessage(roomId: String) {
        roomTask(Current.chatDB.getInstance(), roomId)
    }

    private fun <T : Any> Maybe<T>.doOnUploadError(dao: ChatMessagePendingDao, id: Long): Maybe<T> {
        return doOnError {
            dao.setStatus(ChatMessagePending.STATE_ERR, id)
        }
    }

    private fun ChatDB.sendAndReceiveMessage(roomId: String): Completable {
        val repo = ChatRepository(this)
        val dao = chatMessagePendingDao()
        return Flowable
            .generate<ChatMessagePending> {
                val f = dao.firstUnsentOrSending(roomId).orElseNull()
                if (f == null)
                    it.onComplete()
                else
                    it.onNext(f)
            }
            // todo
            .subscribeOnIO()
            .flatMapCompletable(false, 1) { pending ->
                val id = pending.id
                dao.setStatus(ChatMessagePending.STATE_SENDING, id)

                when (pending.type) {
                    ChatMessagePending.TYPE_IMG_SEND -> Maybe
                        .defer {
                            val m = dao.selectById(id)
                            val f = m?.let { ChatMessagePending.attachmentExt(it) }
                            if (m == null || f == null || !f.file.canRead()) {
                                Maybe.fromAction { repo.deletePendingMessage(pending) }
                            } else fileRepository.msgUploadImage(f.name, f.file.asRequestBody())
                                .map { m to it }
                                .toMaybe()
                        }
                        .observeOnIO()
                        .delayedRetry(2, 2, TimeUnit.SECONDS)
                        .doOnSuccess { v ->
                            val n = ChatMessagePending.imageKeyMessage(
                                v.first,
                                v.second.fileKey
                            )
                            dao.updateOrIgnore(n)
                        }
                        .doOnUploadError(dao, id)
                        .ignoreElement()
                        .onErrorComplete()

                    ChatMessagePending.TYPE_AUDIO_SEND -> Maybe
                        .defer {
                            val m = dao.selectById(id)
                            val f = m?.let { ChatMessagePending.attachmentExt(it) }
                            if (m == null || f == null || !f.file.canRead()) {
                                Maybe.fromAction { repo.deletePendingMessage(pending) }
                            } else fileRepository.msgUploadFile(f.name, f.file.asRequestBody())
                                .map { m to it }
                                .toMaybe()
                        }
                        .observeOnIO()
                        .delayedRetry(1, 2, TimeUnit.SECONDS)
                        .doOnSuccess { v ->
                            val n = ChatMessagePending.audioKeyMessage(
                                v.first,
                                v.second.fileKey
                            )
                            dao.updateOrIgnore(n)
                        }
                        .doOnUploadError(dao, id)
                        .ignoreElement()
                        .onErrorComplete()

                    ChatMessagePending.TYPE_VIDEO_SEND -> Maybe
                        .defer {
                            val m = dao.selectById(id)
                            val f = m?.let { ChatMessagePending.attachmentExt(it) }
                            if (m == null || f == null || !f.file.canRead()) {
                                Maybe.fromAction { repo.deletePendingMessage(pending) }
                            } else fileRepository.msgUploadVideo(f.name, f.file.asRequestBody())
                                .map { Triple(m, it, f) }
                                .toMaybe()
                        }
                        .observeOnIO()
                        .delayedRetry(1, 2, TimeUnit.SECONDS)
                        .doOnSuccess { v ->
                            val (old, uploaded, oldAtt) = v
                            val newAtt = Attachment(
                                uploaded.realFileKey,
                                oldAtt.name,
                                oldAtt.size
                            )
                            val n = ChatMessagePending.videoKeyMessage(
                                old,
                                newAtt,
                                uploaded.key.fileKey
                            )
                            dao.updateOrIgnore(n)
                            UserInfo.cacheFileByKey(Current.userInfo, newAtt.key).let {
                                oldAtt.file.moveOrCopy(it.file)
                                oldAtt.file.delete()
                            }
                        }
                        .doOnUploadError(dao, id)
                        .ignoreElement()
                        .onErrorComplete()

                    ChatMessagePending.TYPE_FILE_SEND -> Maybe
                        .defer {
                            val m = dao.selectById(id)
                            val f = m?.let { ChatMessagePending.attachmentExt(it) }
                            if (m == null || f == null || !f.file.canRead()) {
                                Maybe.fromAction { repo.deletePendingMessage(pending) }
                            } else fileRepository.msgUploadFile2(f.name, f.file.asRequestBody())
                                .map { Triple(m, it, f) }
                                .toMaybe()
                        }
                        .observeOnIO()
                        .delayedRetry(1, 2, TimeUnit.SECONDS)
                        .doOnSuccess { v ->
                            val (old, uploaded, oldAtt) = v
                            val newAtt = Attachment(
                                uploaded.realFileKey,
                                oldAtt.name,
                                oldAtt.size
                            )
                            val n = ChatMessagePending.fileKeyMessage(
                                old,
                                newAtt,
                                uploaded.key.fileKey
                            )
                            dao.updateOrIgnore(n)
                            UserInfo.cacheFileByKey(Current.userInfo, newAtt.key).let {
                                oldAtt.file.moveOrCopy(it.file)
                                oldAtt.file.delete()
                            }
                        }
                        .doOnUploadError(dao, id)
                        .ignoreElement()
                        .onErrorComplete()

                    else -> Maybe
                        .defer {
                            val m = dao.selectById(id)
                            if (m == null) Maybe.fromAction { repo.deletePendingMessage(pending) }
                            else Current.chatApi.sendMessage(m).toState().toMaybe()
                        }
                        .observeOnIO()
                        .doOnDispose {
                            dao.setStatus(ChatMessagePending.STATE_NORMAL, id)
                        }
                        .delayedRetry(2, 2, TimeUnit.SECONDS)
                        .doOnSuccess { v ->
                            dao.setStatusAndTargetId(
                                ChatMessagePending.STATE_SENT,
                                v.orElseNull(),
                                id
                            )
                        }
                        .doOnError {
                            dao.setStatusAndTargetId(
                                ChatMessagePending.STATE_ERR,
                                null,
                                id
                            )
                        }
                        .ignoreElement()
                        .onErrorComplete()
                }
            }
            .andThen(Completable.defer {
                repo.loadBottomMessage(roomId, 80).ignoreElement()
            })
    }

    override fun isDisposed(): Boolean {
        return dc.isDisposed
    }

    override fun dispose() {
        dc.dispose()
    }
}

File: LogiOS/TrustarApp/Chat/MessageTask.swift
//
//  MessageTask.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/11/29.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation
import Combine

class MessageTask: ObservableObject {
    private var bag = Set<AnyCancellable>()
    private var fileRepository = Current.Shared.chatFileRepository
    private var repo = Current.Shared.chatRepository
    private var sessionQueue: DispatchQueue!

    private var tasks: [String] = []
    @Published private var roomId: (ChatDb, String)?

    init() {
        let db = Current.Shared.chatDb
        
        sessionQueue = DispatchQueue(label: "session MessageTask")
        
        db?.chatMessagePendingDao?.unsentOrSendingRoomIds()
            .delay(for: .seconds(0.5), scheduler: sessionQueue)
            .receive(on: sessionQueue)
            .sink(receiveCompletion: { completion in },
                  receiveValue: {[self] ids in
                for id in ids {
                    if db != nil {
                        if tasks.contains(where: { $0 == id.id }) { return }
                        tasks.append(id.id)
                        roomId = (db!, id.id)
                    }
                }
            }).store(in: &bag)
        
        $roomId
            .compactMap({ $0 })
            .setFailureType(to: Never.self)
            .flatMap({ [self] (db, roomId) -> AnyPublisher<(String, StateResult), Never> in
                let f = db.chatMessagePendingDao?.firstUnsentOrSending(roomId: roomId)
                if f == nil {
                    return Just((roomId, StateResult(result: true, message: ""))).eraseToAnyPublisher()
                }
                
                let dao = db.chatMessagePendingDao
                
                let pending = f!
                let id = pending.id!
                dao?.setStatus(status: .STATE_SENDING, id: id)
                
                switch pending.getType() {
                    
                case .TYPE_IMG_SEND:
                    let m = dao?.selectById(id: id)
                    let f: AttachmentExt? = m.map({ $0.attachmentExt(message: $0) }) ?? nil
                    if m == nil || f == nil || !f!.file.canRead() {
                        repo?.deletePendingMessage(pending: pending)
                        return Just((roomId, StateResult(result: true, message: "")))
                            .eraseToAnyPublisher()
                    }
                    
                    return fileRepository!.uploadImage(filename: f!.name, body: f!.contentUri())
                        .replaceError(with: FileKey(fileKey: ""))
                        .eraseToAnyPublisher()
                        .flatMap({ file -> AnyPublisher<(String, StateResult), Never> in
                            if file.fileKey.isEmpty {
                                dao?.setStatus(status: .STATE_ERR, id: id)
                                return Just((roomId, StateResult(result: false, message: "")))
                                    .eraseToAnyPublisher()
                            }
                            
                            let n = ChatMessagePending().imageKeyMessage(m: m!, key: file.fileKey)
                            do {
                                try db.instanceDb?.inDatabase({ db in
                                    try n.update(db)
                                })
                            } catch {
                                debugPrint(error)
                            }
                            
                            _ = f!.file.delete()
                            
                            return Just((roomId, StateResult(result: true, message: "")))
                                .eraseToAnyPublisher()
                        })
                        .eraseToAnyPublisher()
                    
                case .TYPE_AUDIO_SEND:
                    let m = dao?.selectById(id: id)
                    let f: AttachmentExt? = m.map({ $0.attachmentExt(message: $0) }) ?? nil
                    if m == nil || f == nil || !f!.file.canRead() {
                        repo?.deletePendingMessage(pending: pending)
                        return Just((roomId, StateResult(result: true, message: "")))
                            .eraseToAnyPublisher()
                    }
                    
                    return fileRepository!.uploadFile(filename: f!.name, body: f!.contentUri())
                        .replaceError(with: FileKey(fileKey: ""))
                        .eraseToAnyPublisher()
                        .flatMap({ file -> AnyPublisher<(String, StateResult), Never> in
                            if file.fileKey.isEmpty {
                                dao?.setStatus(status: .STATE_ERR, id: id)
                                return Just((roomId, StateResult(result: false, message: "")))
                                    .eraseToAnyPublisher()
                            }
                            
                            let n = ChatMessagePending().audioKeyMessage(m: m!, key: file.fileKey)
                            do {
                                try db.instanceDb?.inDatabase({ db in
                                    try n.update(db)
                                })
                            } catch {
                                debugPrint(error)
                            }
                            
                            _ = f!.file.delete()
                            
                            return Just((roomId, StateResult(result: true, message: "")))
                                .eraseToAnyPublisher()
                        })
                        .eraseToAnyPublisher()

                case .TYPE_VIDEO_SEND:
                    let m = dao?.selectById(id: id)
                    let f: AttachmentExt? = m.map({ $0.attachmentExt(message: $0) }) ?? nil
                    if m == nil || f == nil || !f!.file.canRead() {
                        repo?.deletePendingMessage(pending: pending)
                        return Just((roomId, StateResult(result: true, message: "")))
                            .eraseToAnyPublisher()
                    }
                    
                    return fileRepository!.uploadVideo(filename: f!.name, body: f!.contentUri())
                        .replaceError(with: FileKey2(fileKey: ""))
                        .eraseToAnyPublisher()
                        .flatMap({ file -> AnyPublisher<(String, StateResult), Never> in
                            if file.fileKey.isEmpty {
                                dao?.setStatus(status: .STATE_ERR, id: id)
                                return Just((roomId, StateResult(result: false, message: "")))
                                    .eraseToAnyPublisher()
                            }
                            
                            let old = m!
                            let uploaded = file
                            let oldAtt = f!
                            let newAtt = Attachment(key: uploaded.realFileKey(),
                                                    name: oldAtt.name,
                                                    size: oldAtt.size)
                            let n = ChatMessagePending()
                                .videoKeyMessage(m: old,
                                                cacheFile: newAtt,
                                                key: uploaded.key().fileKey)
                            do {
                                try db.instanceDb?.inDatabase({ db in
                                    try n.update(db)
                                })
                            } catch {
                                debugPrint(error)
                            }
                            
                            let c = Current.Shared.userInfo().cacheByKey(fileKey: newAtt.key, isFile: false)
                            _ = oldAtt.contentUri().moveOrCopy(to: c.file.appendingPathComponent(oldAtt.name, isDirectory: false))
                            _ = oldAtt.file.delete()
                            
                            return Just((roomId, StateResult(result: true, message: "")))
                                .eraseToAnyPublisher()
                        })
                        .eraseToAnyPublisher()
                case .TYPE_FILE_SEND:
                    let m = dao?.selectById(id: id)
                    let f: AttachmentExt? = m.map({ $0.attachmentExt(message: $0) }) ?? nil
                    if m == nil || f == nil || !f!.file.canRead() {
                        repo?.deletePendingMessage(pending: pending)
                        return Just((roomId, StateResult(result: true, message: "")))
                            .eraseToAnyPublisher()
                    }
                    
                    return fileRepository!.uploadFile2(filename: f!.name, body: f!.contentUri())
                        .replaceError(with: FileKey2(fileKey: ""))
                        .eraseToAnyPublisher()
                        .flatMap({ file -> AnyPublisher<(String, StateResult), Never> in
                            if file.fileKey.isEmpty {
                                dao?.setStatus(status: .STATE_ERR, id: id)
                                return Just((roomId, StateResult(result: false, message: "")))
                                    .eraseToAnyPublisher()
                            }
                            
                            let old = m!
                            let uploaded = file
                            let oldAtt = f!
                            let newAtt = Attachment(key: uploaded.realFileKey(),
                                                    name: oldAtt.name,
                                                    size: oldAtt.size)
                            let n = ChatMessagePending()
                                .fileKeyMessage(m: old,
                                                cacheFile: newAtt,
                                                key: uploaded.key().fileKey)
                            do {
                                try db.instanceDb?.inDatabase({ db in
                                    try n.update(db)
                                })
                            } catch {
                                debugPrint(error)
                            }
                            
                            let c = Current.Shared.userInfo().cacheByKey(fileKey: newAtt.key, isFile: false)
                            _ = oldAtt.contentUri().moveOrCopy(to: c.file.appendingPathComponent(oldAtt.name, isDirectory: false))
                            _ = oldAtt.file.delete()
                            
                            return Just((roomId, StateResult(result: true, message: "")))
                                .eraseToAnyPublisher()
                        })
                        .eraseToAnyPublisher()
                    
                default:
                    // Text
                    let m = dao?.selectById(id: id)
                    if  m == nil {
                        repo?.deletePendingMessage(pending: pending)
                        return Just((roomId, StateResult(result: true, message: "")))
                            .eraseToAnyPublisher()
                    }
                    
                    return Current.Shared.chatApi.sendMessage(body: m!, apiKey: Current.Shared.loggedUser?.token ?? "")
                        .replaceError(with: StateResult(result: false, message: ""))
                        .flatMap({ rs -> AnyPublisher<(String, StateResult), Never> in
                            if rs.result {
                                dao?.setStatus(status: .STATE_SENT, id: id)
                            } else {
                                dao?.setStatus(status: .STATE_ERR, id: id)
                            }
                            
                            return Just((roomId, rs))
                                .eraseToAnyPublisher()
                        }).eraseToAnyPublisher()
                }
            })
            .flatMap({ [self] id, state -> AnyPublisher<(String, StateResult), Never> in
                if repo == nil {
                    return Just(("", StateResult(result: true, message: "")))
                        .eraseToAnyPublisher()
                }
                
                return repo!.loadBottomMessage(roomId: id, pageSize: 80)
                    .replaceError(with: .Error(error: .unknow))
                    .map({ e -> AnyPublisher<(String, StateResult), Never> in
                        return Just((id, StateResult(result: true, message: "")))
                            .eraseToAnyPublisher()
                    })
                    .switchToLatest()
                    .eraseToAnyPublisher()
            })
            .receive(on: sessionQueue)
            .sink(receiveValue: { id, state in
                if state.result {
                    self.tasks.removeAll(where: { $0 == id })
                }
            }).store(in: &bag)
    }
    
    func dispose() {
        bag.forEach { $0.cancel()}
    }
}
